diff --git a/sound/soc/qcom/x1e80100.c b/sound/soc/qcom/x1e80100.c
index c81df41ace..2d02f76b77 100644
--- a/sound/soc/qcom/x1e80100.c
+++ b/sound/soc/qcom/x1e80100.c
@@ -109,8 +109,8 @@ static int x1e80100_snd_hw_map_channels(unsigned int *ch_map, int num)
 		break;
 	case 4:
 		ch_map[0] = PCM_CHANNEL_FL;
-		ch_map[1] = PCM_CHANNEL_LB;
-		ch_map[2] = PCM_CHANNEL_FR;
+		ch_map[1] = PCM_CHANNEL_FR;
+		ch_map[2] = PCM_CHANNEL_LB;
 		ch_map[3] = PCM_CHANNEL_RB;
 		break;
 	default:
